lib/eventsource/eventsource.js

```diff
@@ -0,0 +1,326 @@
+'use strict'
+
+const { pipeline } = require('node:stream')
+const {
+  kEvents,
+  kState
+} = require('./symbols')
+const { getGlobalOrigin } = require('../fetch/global')
+const { webidl } = require('../fetch/webidl')
+const { CONNECTING, OPEN, CLOSED, mimeType, defaultReconnectionTime } = require('./constants')
+const { EventSourceStream } = require('./eventsource-stream')
+
+/**
+ * @typedef {object} EventSourceInit
+ * @property {boolean} [withCredentials] indicates whether the request
+ * should include credentials.
+ */
+
+/**
+ * The EventSource interface is used to receive server-sent events. It
+ * connects to a server over HTTP and receives events in text/event-stream
+ * format without closing the connection.
+ * @extends {EventTarget}
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/EventSource
+ * @api public
+ */
+class EventSource extends EventTarget {
+  #url = null
+  #withCredentials = false
+  #readyState = CONNECTING
+  #lastEventId = ''
+  #connection = null
+  #reconnectionTimer = null
+  #controller = new AbortController()
+
+  /**
+   * Creates a new EventSource object.
+   * @param {string} url
+   * @param {EventSourceInit} [eventSourceInitDict]
+   */
+  constructor (url, eventSourceInitDict) {
+    super()
+
+    webidl.argumentLengthCheck(arguments, 1, { header: 'EventSource constructor' })
+
+    // 1. Let baseURL be this's relevant settings object's API base URL.
+    const baseURL = getGlobalOrigin()
+
+    // 2. Let urlRecord be the result of applying the URL parser to url with baseURL.
+    let urlRecord
+
+    try {
+      urlRecord = new URL(url, baseURL)
+    } catch (e) {
+      // 3. If urlRecord is failure, then throw a "SyntaxError" DOMException.
+      throw new DOMException(e, 'SyntaxError')
```

Comment:
In browsers this technically can't fail here so we can use any error we'd like. We should probably match Deno here, do you know what they're doing when URL parsing fails?

Reply:
If I remember correctly SyntaxError is expected from a wpt test.

Reply:
I can't see how this would throw an error in browsers 🤔

Reply:
e.g. new URL('https://_') could result in an error of URL constructor.

Reply:
also i copied it from the websocket implementation

Reply:
yeah, wasn't blocking, was mostly curious - Deno is throwing a SyntaxError DOMException too so it's good 👍 https://github.com/denoland/deno/blob/f45ceb23206b4aaeca123c06bd3d59770f728468/ext/fetch/27_eventsource.js#L178